<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
 <head>
  <meta http-equiv="content-type" content="text/html; CHARSET=gb2312">
  <title>PHP 7 错误处理</title>

  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <link media="all" rel="stylesheet" type="text/css" href="style.css"/>
 </head>
 <body class="docs" class="docs"><div class="navbar navbar-fixed-top">
  <div class="navbar-inner clearfix">
    <ul class="nav" style="width: 100%">
      <li style="float: left;"><a href="language.errors.basics.html">? Basics</a></li>
      <li style="float: right;"><a href="language.exceptions.html">异常处理 ?</a></li>
    </ul>
  </div>
</div>
<div id="breadcrumbs" class="clearfix">
  <ul class="breadcrumbs-container">
    <li><a href="index.html">PHP Manual</a></li>
    <li><a href="language.errors.html">Errors</a></li>
    <li>PHP 7 错误处理</li>
  </ul>
</div>
<div id="layout">
  <div id="layout-content"><div id="language.errors.php7" class="sect1">
 <h2 class="title">PHP 7 错误处理</h2>

 <p class="para">
  PHP 7 改变了大多数错误的报告方式。不同于传统（PHP 5）的错误报告机制，现在大多数错误被作为
  <strong class="classname">Error</strong> 异常抛出。
 </p>

 <p class="para">
  这种 <strong class="classname">Error</strong> 异常可以像 <a href="class.exception.html" class="classname">Exception</a>
  异常一样被第一个匹配的 <em>try</em> / <em>catch</em>
  块所捕获。如果没有匹配的
  <a href="language.exceptions.html#language.exceptions.catch" class="link"><em>catch</em></a>
  块，则调用异常处理函数（事先通过 <span class="function"><a href="function.set-exception-handler.html" class="function">set_exception_handler()</a></span> 注册）进行处理。
  如果尚未注册异常处理函数，则按照传统方式处理：被报告为一个致命错误（Fatal Error）。
 </p>

 <p class="para">
  <strong class="classname">Error</strong> 类并非继承自 <a href="class.exception.html" class="classname">Exception</a>
  类，所以不能用 <code class="code">catch (Exception $e) { ... }</code> 来捕获
  <strong class="classname">Error</strong>。你可以用
  <code class="code">catch (Error $e) { ... }</code>，或者通过注册异常处理函数（
  <span class="function"><a href="function.set-exception-handler.html" class="function">set_exception_handler()</a></span>）来捕获 <strong class="classname">Error</strong>。
 </p>

 <div class="sect2" id="language.errors.php7.hierarchy">
  <h3 class="title"><strong class="classname">Error</strong> 层次结构</h3>

  <ul class="itemizedlist">
   <li class="listitem">
    <span class="simpara"><strong class="classname">Throwable</strong></span>
    <ul class="itemizedlist">
     <li class="listitem">
      <span class="simpara"><strong class="classname">Error</strong></span>
      <ul class="itemizedlist">
       <li class="listitem">
        <span class="simpara"><strong class="classname">ArithmeticError</strong></span>
        <ul class="itemizedlist">
         <li class="listitem">
          <span class="simpara"><strong class="classname">DivisionByZeroError</strong></span>
         </li>
        </ul>
       </li>
       <li class="listitem">
        <span class="simpara"><strong class="classname">AssertionError</strong></span>
       </li>
       <li class="listitem">
        <span class="simpara"><strong class="classname">ParseError</strong></span>
       </li>
       <li class="listitem">
        <span class="simpara"><strong class="classname">TypeError</strong></span>
       </li>
      </ul>
     </li>
     <li class="listitem">
      <span class="simpara"><a href="class.exception.html" class="classname">Exception</a></span>
      <ul class="itemizedlist">
       <li class="listitem">
        <span class="simpara">...</span>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </div>
</div><div id="usernotes">
 <div class="head">
  <h3 class="title">User Contributed Notes</h3>
 </div>
<div id="allnotes">  <a name="118610""></a>
  <div class="note">
   <strong class="user">hungry dot rahly at gmail dot com</strong>
   <a href="#118610" class="date">08-Jan-2016 09:26</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
You can catch both exceptions and errors by catching(Throwable)</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
</div>
</div>
</div></div></body></html>
